src/Rules.Framework/ConditionTypeExtractor.cs

```diff
@@ -0,0 +1,76 @@
+namespace Rules.Framework
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Rules.Framework.Core;
+    using Rules.Framework.Core.ConditionNodes;
+
+    /// <summary>
+    /// Extracts Conditions Types from a Group of Rules.
+    /// </summary>
+    /// <typeparam name="TContentType">The content type that allows to categorize rules.</typeparam>
+    /// <typeparam name="TConditionType">
+    /// The condition type that allows to filter rules based on a set of conditions.
+    /// </typeparam>
+    public class ConditionTypeExtractor<TContentType, TConditionType> : IConditionTypeExtractor<TContentType, TConditionType>
+    {
+        /// <summary>
+        /// Get the unique condition types associated with rules of a specific content type/>.
+        /// </summary>
+        /// <param name="matchedRules"></param>
+        /// <remarks>
+        /// <para>
+        /// A set of rules is requested to rules data source and all conditions are evaluated
+        /// against them to provide a set of matches.
+        /// </para>
+        /// <para>All rules matching supplied conditions are returned.</para>
+        /// </remarks>
+        /// <returns>the matched rule; otherwise, null.</returns>
+        public IEnumerable<TConditionType> GetConditionTypes(IEnumerable<Rule<TContentType, TConditionType>> matchedRules)
+        {
+            var conditionTypes = new List<TConditionType>();
+
+            if (!matchedRules.Any())
+            {
+                return conditionTypes;
+            }
+
+            foreach (var rule in matchedRules)
+            {
+                var rootCondition = rule.RootCondition;
+
+                VisitConditionNode(rootCondition, conditionTypes);
```

Comment:
I think you'll have to check if `rootCondition` is not null. If I'm not mistaken, `VisitConditionNode(...)` will fail with a `NotSupportedException` if you provide a rule without a `RootCondition` (null).

Can you please add a unit test to confirm my suspicion, and ensure that we have that scenario covered and resolved?

Reply:
Yes, you are correct, such is the nature of the switch case in that method. I've added a try catch to the method to deal with those cases. Do you agree with the approach?